### ISSUE TITLE
Does this work for LLaMA models?

### ISSUE THREAD

**Issue by marcoripa96** [hyperonym/basaran#57]
I wanna stream the output of LLaMA models is it possible using this?

**Comment by peakji** [comment]
The [implementation of LLaMA](https://github.com/huggingface/transformers/issues/21796) has just been merged into HF transformers, but it seems that it did not make it to the latest release.

Basaran will support LLaMA as soon as it is included in HF transformers.

**Comment by deep-diver** [comment]
I am building a Gradio app with LLaMA, can I borrow your codebase of token generator?

**Comment by peakji** [comment]
> I am building a Gradio app with LLaMA, can I borrow your codebase of token generator?

Yes, of course! Basaran is released under the MIT License.

> I have checked StreamModel works nicely with LLaMA...

If that's the case, do you think it's necessary for us to release Basaran on PyPI for developers to use it as a library? This would be more maintainable than copying and pasting.

Also, which LLaMA model are you testing? Could you share the link to the model on the Hugging Face Hub so that we can test it as well? Thanks!

**Comment by deep-diver** [comment]
I have fine-tuned 30B version of Alpaca-LoRA! I am currently running a demo publicly

https://notebooksf.jarvislabs.ai/43j3x9FSS8Tg0sqvMlDgKPo9vsoSTTKRsX4RIdC3tNd6qeQ6ktlA0tyWRAR3fe_l/

And here is the repo for this app
: https://github.com/deep-diver/Alpaca-LoRA-Serve

Super thanks ! Yeay separate packages would be wonderful!!

**Comment by marcoripa96** [comment]
I was about to ask the same thing: to provide a different package with the streaming classes and helpers.

@deep-diver did you managed to make it work with alpaca-lora? I tried (using alpaca-lora 7B) by copying and pasting classes and helpers of basaran and I always ended up with an error caused by the attention mask tensor shape being different from what was required during inference.
Do you have the code where you tried that? Thanks a lot!

**Comment by deep-diver** [comment]
Check this out. I will share when it is done. Some minor issues are remained

https://twitter.com/algo_diver/status/1638079375085305856

**Comment by peakji** [comment]
Basaran is now available as a library on TestPyPI:

```bash
pip install -i https://test.pypi.org/simple/ basaran
```

The PyPI release will be part of the next release!

**Comment by peakji** [comment]
Basaran is now available as a library on [PyPI](https://pypi.org/project/basaran/). To use it programmatically, install it with `pip`:

```bash
pip install basaran
```

Use the `load_model` function to load the specified model and generate streaming output by calling the model:

```python
from basaran.model import load_model

model = load_model("user/repo")

for choice in model("once upon a time"):
    print(choice)
```

The [examples](https://github.com/hyperonym/basaran/tree/master/examples) directory contains examples of [using Basaran as a library](https://github.com/hyperonym/basaran/blob/master/examples/basaran-python-library/main.py).

**Comment by peakji** [comment]
@marcoripa96 The error seems to be related to the model it self. Can you run the model without using Basaran?

AFAIK, LLaMA support in HF Transformer is still in active development.

**Comment by deep-diver** [comment]
check out the updated demo
: https://notebookse.jarvislabs.ai/BuOu_VbEuUHb09VEVHhfnFq4-PMhBRVCcfHBRCOrq7c4O9GI4dIGoidvNf76UsRL

I am curious how to boost the inference speed up?

**Comment by peakji** [comment]
> I am curious how to boost the inference speed up?

@deep-diver Have you tried quantization? Basaran provides two [quantization options](https://github.com/hyperonym/basaran/blob/master/basaran/model.py#L309), `load_in_8bit` (INT8) and `half_precision` (FP16), but I'm not sure if LLaMA's implementation supports them.

**Comment by deep-diver** [comment]
I am using peft for 8bit since fp16 wouldt let the 30b model fit into 40gb(which is the limit of vram that I have)

**Comment by deep-diver** [comment]
Just curious StreamModel supports all parameters in GenerationConfig?

**Comment by peakji** [comment]
> I am using peft for 8bit since fp16 wouldt let the 30b model fit into 40gb(which is the limit of vram that I have)

Then I guess this is probably the fastest speed possible under the current resource constraints.

> Just curious StreamModel supports all parameters in GenerationConfig?

No. All **model arguments** are supported, but most **generation options** (like `top_k`) are not supported.

**Comment by deep-diver** [comment]
is that because of a sort of technical difficulties? like hard to monkey patch the current implementation of `transformers`?

**Comment by peakji** [comment]
Most generation options require a reimplementation to support streaming output, simply copying code from `transformers` won't work.

Considering that Basaran's goal is to be compatible with the OpenAI API, the common arguments were prioritized for implementation.

**Comment by deep-diver** [comment]
thanks for the clarification. I can see you made a huge contiribution to the open source world with this one. I really want to say thank you! :)

**Comment by deep-diver** [comment]
Have a look at my repo
https://github.com/deep-diver/Alpaca-LoRA-Serve

I also run a couple of demos

**Comment by peakji** [comment]
Basaran v0.15.0 now officially supports LLaMA! Tested with [Enoch/llama-7b-hf](https://huggingface.co/Enoch/llama-7b-hf) ([Dockerfile](https://github.com/hyperonym/basaran/blob/master/deployments/bundle/llama-7b.Dockerfile)) and [zpn/llama-7b](https://huggingface.co/zpn/llama-7b).

Theoretically, it should support any LLaMA-based models, such as Alpaca. Just keep in mind is to avoid outdated models using legacy class names and configs, like [decapoda-research/llama-7b-hf](https://huggingface.co/decapoda-research/llama-7b-hf).

![playground](https://user-images.githubusercontent.com/219051/232123152-4936484e-fcce-4898-bd5b-d2659bbe7ae7.gif)
